src/Access/AccessControlManager.h

```diff
@@ -109,7 +109,7 @@ class AccessControlManager : public MultipleAccessStorage
     bool isSettingNameAllowed(const std::string_view & name) const;
     void checkSettingNameIsAllowed(const std::string_view & name) const;
 
-    UUID login(const String & user_name, const String & password, const Poco::Net::IPAddress & address) const;
+    UUID login(const Credentials & credentials, const Poco::Net::IPAddress & address) const;
```

Comment:
Maybe we can wrap `address` into the credentials too. And it seems to be logical to make one more overload for `Context::setUser()`, which will get credentials instead of `user_name`+`password`, like the following:

```
void Context::setCredentials(const Credentials & credentials)
{
   // main function, does authentication and get user's name (either from credentials or from the authentication process)
}

void Context::setUser(const String & user_name, const String & password)
{
  /// wrap user_name and password into Credentials and then call setCredentials
}

// In this way we could also get rid of Context::setUserWithoutCheckingPassword()
```

Reply:
Technically, the address is not a part of credentials. I.e., if credentials are used or stored anywhere beyond the session authentication pipeline, the address would be irrelevant.

> And it seems to be logical to make one more overload for Context::setUser()

It exists.

Reply:
ok, but the name `setCredentials()` seems to be more appropriate in this case.

Reply:
I thought about that, but it seems we actually set the user in the context there, not credentials. We check the credentials and extract the user from them.

Reply:
> > 180 commits are a lot. If you can squash them it would be nice.
> 
> Is **Allow squash merging** enabled for the repo?
> If yes, you can choose to squash during the merge, actually.
> If not, I will squash manually.

I can allow squashing when I merge it. But it's always difficult not to forget about it when it's time to merge. I prefer to squash my own PR because in my experience it makes merging with master easier. And it lessens the chances of merging errors like you have here with brotli.